core/src/main/java/org/sonarsource/sonarlint/core/SonarLintClientImpl.java

```diff
@@ -94,16 +96,35 @@ private void checkStarted() {
 
   private static RuntimeException handleException(RuntimeException t) {
     if (LOG.isDebugEnabled()) {
-      return t;
+      // In DEBUG mode always return full exception
+      return convertToSonarLintException(t);
     }
 
     for (Throwable y : Throwables.getCausalChain(t)) {
       if (y instanceof MessageException) {
-        return (MessageException) y;
+        return convertToSonarLintException(y);
       }
     }
 
-    return t;
+    return convertToSonarLintException(t);
+  }
+
+  private static RuntimeException convertToSonarLintException(@Nullable Throwable t) {
+    if (t == null) {
+      return null;
+    }
+    Throwable cause = convertToSonarLintException(t.getCause());
+    if (t instanceof RuntimeException && isPartOfJDK(t)) {
+      return (RuntimeException) t;
+    }
+    // If exception is not part of the API (and not part of JDK) then we should convert it to a SonarLintException
+    SonarLintException sonarLintException = new SonarLintException(t.toString(), t.getMessage(), cause);
+    sonarLintException.setStackTrace(t.getStackTrace());
+    return sonarLintException;
+  }
+
+  private static boolean isPartOfJDK(Throwable t) {
+    return t.getClass().getClassLoader() == "".getClass().getClassLoader();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)